### PR TITLE
Initial Threadservice

### PR DESCRIPTION
**Pull request by sanderpick**
This goes along with recent work here: https://github.com/textileio/go-textile-threads

### CODE REVIEW

File: models/models.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+	// Peers holds Roles for specific peers.
+	Peers() map[peer.ID]Role
+}
+
+// ACL defines the access roles for a thread and its docs.
+type ACL interface {
+	Roles
+	Docs() map[string]Roles
+}
```
Comment (sanderpick) [inline models/models.go:48]: Stubbing in some models here @carsonfarmer just to retain these ideas... though we may not end up leveraging these in this form.

File: thread/event.go [hunk 2]
```diff
@@ -1,23 +1,39 @@
...
 	format.Node
 
-	Header() EventHeader
-	Body() format.Node
+	// HeaderID returns the cid of the event header.
+	HeaderID() cid.Cid
+
+	// GetHeader loads and optionally decrypts the event header.
+	// If no key is given, the header time and key methods will return an error.
+	GetHeader(context.Context, format.DAGService, crypto.DecryptionKey) (EventHeader, error)
```
Comment (sanderpick) [inline thread/event.go:21]: Linked objects like header here are all lazy loaded.

File: threadservice/options.go [hunk 3]
```diff
@@ -1,30 +1,100 @@
...
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/textileio/go-textile-core/crypto"
 	"github.com/textileio/go-textile-core/thread"
 )
 
-// PutOpt is an instance helper for creating options
+// AddOpt is an instance helper for creating add options.
+var AddOpt AddOption
```
Comment (sanderpick) [inline threadservice/options.go:14]: Add, Put, and Pull options for dealing with the API.

File: thread/event.go [hunk 4]
```diff
@@ -1,23 +1,39 @@
...
 type Event interface {
 	format.Node
 
-	Header() EventHeader
-	Body() format.Node
+	// HeaderID returns the cid of the event header.
+	HeaderID() cid.Cid
+
+	// GetHeader loads and optionally decrypts the event header.
+	// If no key is given, the header time and key methods will return an error.
```
Comment (andrewxhill) [inline thread/event.go:20]: So is that saying, keys are mandatory?

Comment (sanderpick) [reply thread/event.go:20]: currently, yeah - some more thought needed around making keys optional

**Review by andrewxhill — APPROVED** [review]

**Pull request merged**